geo/geotransform: implement ST_Transform

otan opened the pull request:
This PR implements ST_Transform, allowing the transformation from one
SRID to another.

The `geoprojbase` package defines a barebones set of types as well as a
hardcoded list of SRIDs to keep in memory. I've only filled in a few for
now, and will save updating this for a later PR.

`geoproj` is strictly a C library interface library which performs the
necessary transformations.

`geotransform` is where the function is actually handled and to be used
by `geo_builtins.go`.

Resolves #49055
Resolves #49056
Resolves #49057
Resolves #49058

Release note (sql change): Implemented the ST_Transform function for
geometries.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49783)
<!-- Reviewable:end -->

sumeerbhola reviewed (changes requested):
> Reviewed 14 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49783)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geoproj/geoproj.go, line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vPfib39sNLgySPklQ:-M8vPfib39sNLgySPklR:b-r9ea0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoproj/geoproj.go#L33)):*
> ```Go
> 
> // goToCSlice returns a CR_PROJ_Slice from a given Go byte slice.
> func goToCSlice(b []byte) C.CR_PROJ_Slice {
> ```

Can we define the string and slice c++ structs and the conversions once, and use them in all the library wrappers we are writing? Or is that tricky in our current build?
___
*[pkg/geo/geoproj/geoproj.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vQgF48fhQXWitMPkp:-M8vQgF48fhQXWitMPkq:bj651i8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoproj/geoproj.go#L73)):*
> ```Go
> 	}
> 	if err := cStatusToUnsafeGoBytes(C.CR_PROJ_Transform(
> 		goToCSlice([]byte(from)),
> ```

isn't `[]byte(from)` a copy? If so, should we change `Proj4Text` to be a `[]byte` and avoid this copy.
___
*[pkg/geo/geoproj/proj.h, line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vLcBZ-ZIvykjOEVrj:-M8vLcBZ-ZIvykjOEVrk:bc7m22y) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoproj/proj.h#L18)):*
> ```objc
> 
> // CR_PROJ_Slice contains data that does not need to be freed.
> // // It can be either a Go or C pointer (which indicates who allocated the
> ```

nit: extra //
___
*[pkg/geo/geoproj/proj.h, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vM6pIBz2kp0ESCJIB:-M8vM6pIBz2kp0ESCJIC:b-ww0606) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoproj/proj.h#L25)):*
> ```objc
> } CR_PROJ_Slice;
> 
> typedef CR_PROJ_Slice CR_PROJ_Status;
> ```

are all the status strings `char*` literals?
___
*[pkg/geo/geoproj/proj.h, line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vMzAw9zxo-vvEiBp1:-M8vMzAw9zxo-vvEiBp2:b-bus8o0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoproj/proj.h#L27)):*
> ```objc
> typedef CR_PROJ_Slice CR_PROJ_Status;
> 
> CR_PROJ_Status CR_PROJ_Transform(CR_PROJ_Slice from, CR_PROJ_Slice to,
> ```

can you add a function comment clarifying that the points are (x[i], y[i], z[i]) for  0 <= i < point_coint

Also, the `from`, `to` here and in the caller `Project` is a bit confusing -- it suggests that `to` is being written to. How about something like `fromSpec`, `toSpec`.
___
*[pkg/geo/geoproj/proj.cc, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vNcBD00W30dAwfZ3C:-M8vNcBD00W30dAwfZ3D:b-kkh8j0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoproj/proj.cc#L20)):*
> ```cpp
> 
> namespace {
> CR_PROJ_Status CR_PROJ_ErrorFromErrorCode(int code) {
> ```

nit: I thought we'd decided stylistically to use `char* err` (same for other similar code in this file) when doing GEOS (geos.h mostly follows that style though there is some inconsistency in recent additions).
___
*[pkg/geo/geoproj/proj.cc, line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vOQgR-8Lb8pyBLuJZ:-M8vOQgR-8Lb8pyBLuJ_:b-1yetcc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoproj/proj.cc#L35)):*
> ```cpp
>   CR_PROJ_Status err = {.data = NULL, .len = 0};
>   auto ctx = pj_ctx_alloc();
>   auto fromPJ = pj_init_plus_ctx(ctx, std::string(from.data, from.len).c_str());
> ```

is this expensive to initialize each time? it needs to parse a structured string at least.

And we could avoid the copy needed for `c_str` by storing null terminated strings in `ProjInfo`
___
*[pkg/geo/geoprojbase/.clang-format, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vK9Iu8VS_56hfLbfa:-M8vK9Iu8VS_56hfLbfb:bv6ift5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoprojbase/.clang-format#L1)):*
> ```YAML
> ---
> ```

where did the contents of this file come from, and where is the C++ code that this is formatting -- I don't see any non-Go files in this directory.
___
*[pkg/geo/geoprojbase/geoprojbase.go, line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vKePDDWmElzApZ8x-:-M8vKePDDWmElzApZ8x0:bueyvjr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoprojbase/geoprojbase.go#L26)):*
> ```Go
> 	AuthName  string
> 	AuthSRID  int
> 	SRText    string
> ```

can you add comments for AuthName, AuthSRID, SRText
___
*[pkg/geo/geoprojbase/projections.go, line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vLPe2-WHr7ldmTI_m:-M8vLPe2-WHr7ldmTI_n:b-kmtgz6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoprojbase/projections.go#L27)):*
> ```Go
> 	},
> 	4004: {
> 		SRID:      4004,
> ```

is 4004 a popular one?
___
*[pkg/geo/geotransform/geotransform.go, line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vUFx48rK4PUZI8baP:-M8vUFx59mnh3lfhticK:bwwzfro) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geotransform/geotransform.go#L108)):*
> ```Go
> 	numCoords := len(flatCoords) / layout.Stride()
> 	xCoords := make([]float64, numCoords)
> 	yCoords := make([]float64, numCoords)
> ```

`numCoords` may be small, so it may be worthwhile doing a single slice allocation and using different parts of it for x, y, z.
___
*[pkg/geo/geotransform/geotransform.go, line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vTuPN6EzZzQwIcznA:-M8vTuPO6hxTYj8mx7wn:b-3xu5e3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geotransform/geotransform.go#L115)):*
> ```Go
> 		yCoords[i] = flatCoords[base+1]
> 		if layout.ZIndex() != -1 {
> 			zCoords[i] = flatCoords[base+layout.ZIndex()]
> ```

// In this case ZIndex() is 2


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49783)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[pkg/geo/geoproj/geoproj.go, line 33 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vPfib39sNLgySPklQ:-M8vYooT6yNQFBwklkTp:b-vt16vp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoproj/geoproj.go#L33)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can we define the string and slice c++ structs and the conversions once, and use them in all the library wrappers we are writing? Or is that tricky in our current build?
</blockquote></details>

i think it's tricky to find the right incantations in our current build and i'd rather avoid it for now.
___
*[pkg/geo/geoproj/geoproj.go, line 73 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vQgF48fhQXWitMPkp:-M8v_IPT5EDf2E_IoKOO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoproj/geoproj.go#L73)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

isn't `[]byte(from)` a copy? If so, should we change `Proj4Text` to be a `[]byte` and avoid this copy.
</blockquote></details>

Done.
___
*[pkg/geo/geoproj/proj.h, line 25 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vM6pIBz2kp0ESCJIB:-M8vbPsx1LUalF6_RXo7:bb77uci) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoproj/proj.h#L25)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

are all the status strings `char*` literals?
</blockquote></details>

Yep!
___
*[pkg/geo/geoproj/proj.h, line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vMzAw9zxo-vvEiBp1:-M8vjVNtAKEJz_CilYyQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoproj/proj.h#L27)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you add a function comment clarifying that the points are (x[i], y[i], z[i]) for  0 <= i < point_coint

Also, the `from`, `to` here and in the caller `Project` is a bit confusing -- it suggests that `to` is being written to. How about something like `fromSpec`, `toSpec`.
</blockquote></details>

Done.
___
*[pkg/geo/geoproj/proj.cc, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vNcBD00W30dAwfZ3C:-M8vbhNmEFBGe_vAwiJP:bn9a9ct) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoproj/proj.cc#L20)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: I thought we'd decided stylistically to use `char* err` (same for other similar code in this file) when doing GEOS (geos.h mostly follows that style though there is some inconsistency in recent additions).
</blockquote></details>

copied the `.clang-format` file to fix this.
___
*[pkg/geo/geoproj/proj.cc, line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vOQgR-8Lb8pyBLuJZ:-M8w-o-o50fd85uTA_b8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoproj/proj.cc#L35)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is this expensive to initialize each time? it needs to parse a structured string at least.

And we could avoid the copy needed for `c_str` by storing null terminated strings in `ProjInfo`
</blockquote></details>

Done.
___
*[pkg/geo/geoprojbase/geoprojbase.go, line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vKePDDWmElzApZ8x-:-M8vjUOX5LqTl2erRTWK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoprojbase/geoprojbase.go#L26)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

can you add comments for AuthName, AuthSRID, SRText
</blockquote></details>

Done.
___
*[pkg/geo/geoprojbase/projections.go, line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vLPe2-WHr7ldmTI_m:-M8vc_779Lg4b6y4DQ91:b-cmwheq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoprojbase/projections.go#L27)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is 4004 a popular one?
</blockquote></details>

no idea, the only popular latlng one i know is 4326. the survey says 3857 was most popular, which wasn't latlng. i picked an arbitrary one.
___
*[pkg/geo/geotransform/geotransform.go, line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vUFx48rK4PUZI8baP:-M8vicPA90JVfO9FLWC2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geotransform/geotransform.go#L108)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

`numCoords` may be small, so it may be worthwhile doing a single slice allocation and using different parts of it for x, y, z.
</blockquote></details>

Done.
___
*[pkg/geo/geotransform/geotransform.go, line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vTuPN6EzZzQwIcznA:-M8vic9S8SX4gXnZtE4x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geotransform/geotransform.go#L115)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

// In this case ZIndex() is 2
</blockquote></details>

Done.
___
*[pkg/geo/geoprojbase/.clang-format, line 1 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vK9Iu8VS_56hfLbfa:-M8vcUTj2e2Cyo-ju6Gn:bbgn5dh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoprojbase/.clang-format#L1)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

where did the contents of this file come from, and where is the C++ code that this is formatting -- I don't see any non-Go files in this directory.
</blockquote></details>

this was in the wrong dir -- moving to ./pkg/geo/geoproj


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 10 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-:-M8zO6qS051DswRHXOr9:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geoproj/proj.h, line 27 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8vMzAw9zxo-vvEiBp1:-M8zAFur5guxpanARbVq:b-eymhk3) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7ed835bd26f409c046624e7ee7997b00506bc5c/pkg/geo/geoproj/proj.h#L27)):*
(reminder)

> can you add a function comment clarifying that the points are (x[i], y[i], z[i]) for 0 <= i < point_coint
___
*[pkg/geo/geoprojbase/geoprojbase.go, line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/49783#-M8zAiyrAkJLIX3EQDWV:-M8zAiyrAkJLIX3EQDWW:b-xtqvl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/67d930d63ea49f73412b254163c8a6372569d321/pkg/geo/geoprojbase/geoprojbase.go#L47)):*
> ```Go
> type ProjInfo struct {
> 	SRID geopb.SRID
> 	// AuthName is authority who has provided this projection (e.g. ESRI, EPSG).
> ```

nit: ... the authority ...


<!-- Sent from Reviewable.io -->

otan committed:
geo/geotransform: implement ST_Transform

This PR implements ST_Transform, allowing the transformation from one
SRID to another.

The `geoprojbase` package defines a barebones set of types as well as a
hardcoded list of SRIDs to keep in memory. I've only filled in a few for
now, and will save updating this for a later PR.

`geoproj` is strictly a C library interface library which performs the
necessary transformations.

`geotransform` is where the function is actually handled and to be used
by `geo_builtins.go`.

Release note (sql change): Implemented the ST_Transform function for
geometries.

otan commented:
bors r=sumeerbhola
